### ISSUE TITLE
`Ltac2 Eval Std.unify` believes that all terms unify

### ISSUE THREAD

**Issue by JasonGross** [coq/coq#18021]
#### Description of the problem

```coq
From Ltac2 Require Import Ltac2.
Ltac2 test () := Std.unify 'True 'False.
Goal True. Fail test (). Abort.
(* The command has indeed failed with message:
Unable to unify "True" with "False".
 *)
Ltac2 Eval test ().
(* success: - : unit = () *)
```
This makes `Ltac2 Eval` effectively useless for complex tactics that invoke unification.

#### Coq Version

8.19 (61ee398ed32f9334dd664ea8ed2697178e6e3844)

**Comment by rlepigre** [comment]
This is due to a combination of two things:
- `Tactics.unify` (used as implementation of Ltac2's `Std.unify`) using `Proofview.Goal.enter` (see [here](https://github.com/coq/coq/blob/master/tactics/tactics.ml#L3374-L3375)),
- `Ltac2 Eval e` runs `e` without any goal existing.

It is a bit surprising that `Unify.unify` does `Proofview.Goal.enter`: it should probably rather do `Proofview.Goal.enter_one`? In any case, the result of that would be that your `Ltac2 Eval` call will fail due to no goal existing.

**Comment by rlepigre** [comment]
Actually, you can fix your problem by adding `Goal True.` before your `Ltac2 Eval`!

**Comment by JasonGross** [comment]
I don't think unify should enter goals at all.  Any focusing should happen during construction of the constr arguments.  Presumably you don't need to focus to get your hands on an evar map, and focusing (or not) should not depend on goals at all.

**Comment by rlepigre** [comment]
Yeah, I guess if there are no goals the unification tactics should fall back to using `tclENV` / `tclSIGMA`. This kind of behaviour is already implemented for some tactics (there was a related discussion in https://github.com/coq/coq/pull/17878, where I plan to provide support for that use-case at the tactic declaration level).

**Comment by SkySkimmer** [comment]
>Presumably you don't need to focus to get your hands on an evar map

You need the env from the right goal to unify.

**Comment by JasonGross** [comment]
> > Presumably you don't need to focus to get your hands on an evar map
> 
> You need the env from the right goal to unify.

Are you saying that the evar map is stored in the goal, and when invoking `unify` you may get a term with evars and you don't know what evar map they are associated to?

**Comment by SkySkimmer** [comment]
No, you need the env because there are variables there

**Comment by JasonGross** [comment]
Ah, I see, that makes sense.  So I guess the behavior should be to use the global env if there are no goals, and otherwise to focus?  (I'm not sure what the behavior should be if there are multiple goals under focus.  I guess we should follow the behavior of `open_constr:(...)` in that case?  I guess maybe we need to focus, for example in `let x := 'y in split > [ clear y | ]; unify x x` or w/e)

**Comment by JasonGross** [comment]
Here is an example where having `unify` only work when there are open goals is untenable because opening a goal is too slow: https://github.com/coq/coq/pull/17899#issuecomment-1711786569

**Comment by SkySkimmer** [comment]
I'm not sure how much we care about cases where the user defined a bunch of horrible `Let`s.

**Comment by JasonGross** [comment]
Can the issue where native compilation cannot be disabled on a per-definition basis be fixed then?

**Comment by JasonGross** [comment]
Oh, I see you have done that in https://github.com/coq/coq/pull/18033!

**Comment by SkySkimmer** [comment]
So what do we do with this issue?

**Comment by JasonGross** [comment]
I would still like to see the focus-or-fallback-to-global a la `open_constr:(...)` implemented
